cassiopeia/data/CassiopeiaTree.py

```diff
@@ -557,6 +557,40 @@ def set_time(self, node: str, new_time: float) -> None:
                 self.get_time(child) - new_time
             )
 
+    def set_times(self, time_dict: Dict[str, float]) -> None:
+        """Sets the time of all nodes in the tree.
+
+        Importantly, this maintains consistency with the rest of the tree. In
+        other words, setting the time of all nodes will change the length of
+        the edges too. This function requires monotonicity of times are
+        maintained (i.e. no negative branch lengths).
+
+        Args:
+            time_dict: Dictionary mapping nodes to their time.
+
+        Raises:
+            CassiopeiaTreeError if the tree is not initialized, or if the time
+            of any parent is greater than that of a child.
+        """
+        self.__check_network_initialized()
+
+        # TODO: Check that the keys of time_dict match exactly the nodes in the
+        # tree and raise otherwise?
+        # Currently, if nodes are missing in time_dict, code below blows up. If
+        # extra nodes are present, they are ignored.
+
+        for (parent, child) in self.edges:
+            time_parent = time_dict[parent]
+            time_child = time_dict[child]
+            if time_parent > time_child:
+                raise CassiopeiaTreeError(
+                    "Time of parent greater than that of child: "
+                    f"{time_parent} > {time_child}"
+                )
+            self.__network[parent][child]["length"] = time_child - time_parent
+        for node, time in time_dict.items():
+            self.__network.nodes[node]["time"] = time
+
```

Comment:
> I think if you were to traverse the tree correctly while updating, you should avoid any exceptions being raised.

This is exactly why I decided not to use `set_time`: we would have to do some juggling to avoid exceptions. We can think about this though, I agree having the logic in only one place would be ideal. Another hacky solution: we can have a `pedantic` boolean flag in the `set_time` method, which by default is `True`, but when set to `False` will not enforce any constraints. It can be hidden in a private method such that it is not exposed to users.

Reply:
Ah, I sort of like the `pedantic` solution. 

Another solution -- you could do a bottom-up pass to begin with setting all times to be `np.inf`, and then do a top-down pass setting the times to those in the `time_dict`. I think this is guaranteed to not break anything, right?